sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -876,7 +885,30 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def join(right: Dataset[_]): DataFrame = withPlan {
-    Join(logicalPlan, right.logicalPlan, joinType = Inner, None, JoinHint.NONE)
+    val (joinLeft, joinRight) = prepareJoinPlan(this, right)
+    Join(joinLeft, joinRight, joinType = Inner, None, JoinHint.NONE)
+  }
+
+  // Called by `Dataset#join`, to attach the Dataset id to the logical plan, so that we
+  // can resolve column reference correctly later. See `ResolveDatasetColumnReference`.
+  private def createPlanWithDatasetId(): LogicalPlan = {
+    if (!sparkSession.sessionState.conf.getConf(SQLConf.RESOLVE_DATASET_COLUMN_REFERENCE)) {
+      return logicalPlan
+    }
+
+    // The alias should start with `SubqueryAlias.HIDDEN_ALIAS_PREFIX`, so that `SubqueryAlias` can
```

Comment:
I am not a big fan of using special strings as markers, I'd rather introduce a new attribute to carry on the information.

Reply:
It's at plan level, we can't use attributes to carry information here.

Reply:
I meant adding a new attribute to the `SubqueryPlan` case class, why isn't it possible?

Reply:
It's possible, but I'm a little worried about changing the constructor of a widely used plan node like `SubqueryAlais`. It's also a little weird to put the dataset concept into the catalyst module.

Another idea I've thought about is adding a new no-op plan node to carry the dataset id. I gave it up because no-op node can be troublesome according to the past experience with `ResolvedHint` and `AnalysisBarrier`.

Reply:
> Another idea I've thought about is adding a new no-op plan node to carry the dataset id

Yes, that's exactly the reason why I'd like to avoid adding this node. I think I'd prefer a no-op node if this is really needed actually, since this is used also in other places and we may introduce side-effects using it in a different way from what it is intended to do now.

Reply:
@cloud-fan I was thinking: in order to avoid adding another "placeholder" plan, since here we are dealing only with joins, what about adding a `leftDatasetId` and `rightDatasetId` to the join operator?

Reply:
This works, but I'd rather hack into `SubqueryAlias` instead of `Join`, as `Join` node is more widely used...

Reply:
and things can get tricky if join reorder happens.

Reply:
> and things can get tricky if join reorder happens.

yes, I also thought of this, but still I am not very confident about using `SubqueryAlias`: that node is meant for a different purpose and I think ew already have things used in a hacky way and this makes more error prone any future change, because we need to think not only to the logic but also we need to remember all the "hacks" done with it. Honestly rather than using a `SubqueryAlias`, if no other option works, I'd prefer adding a new node (maybe a sublass of it if we need common functionalities).

Reply:
IMO, `SubqueryAlias` might be safer than new node. Because it is well known to Spark SQL community. A new node which isn't aware by others is more dangerous.